flocker/node/functional/test_docker.py

```diff
@@ -170,6 +174,97 @@ def started(_):
         d.addCallback(started)
         return d
 
+    @require_docker_version(
+        '1.6.0',
+        'This test uses the registry:2 image '
+        'which requires Docker-1.6.0 or newer. '
+        'See https://docs.docker.com/registry/deploying/ for details.'
+    )
+    def test_private_registry_image(self):
+        """
+        ``DockerClient.add`` can start containers based on an image from a
+        private registry.
+
+        A private registry is started in a container according to the
+        instructions at:
+         * https://docs.docker.com/registry/deploying/
+
+        An image is pushed to that private registry and then a Flocker
+        application is started that uses that private repository image name.
+
+        Docker can pull from a private registry without any TLS configuration
+        as long as it's running on the local host.
+        """
+        registry_name = random_name(self)
+        registry_port = find_free_port()[1]
+        repository = '127.0.0.1:{}'.format(registry_port)
+        # Total image name seems to be limited to 30 bytes for some versions of
+        # Docker: https://github.com/docker/docker/issues/10392
+        name = random_name(self).lower()[-(29 - len(repository)):]
```

Comment:
I _think_ this would be simpler and more correct if we did the truncation when we construct the `DockerImage` below, because it could be expressed as `repository=(repository + '/' + name)[:30]`.